File: leetcode/Validate Binary Search Tree.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return a boolean
    def isValidBST(self, root):
        if root:
            self.inodr = []
            self.inorder(root)
            node = self.inodr[0]
            for i in self.inodr[1:]:
                if node >= i:
                    return False
                else:
                    node = i
        return True
    
    def inorder(self, root):
        if root:
            self.inorder(root.left)
            self.inodr.append(root.val)
            self.inorder(root.right)

File: leetcode/cpp/Validate_Binary_Search_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        vector<int> assist;
        return isValidBSTHelper(root, assist);
    }
    
    bool isValidBSTHelper(TreeNode *root, vector<int> &a) {
        if (root) {
            bool left = isValidBSTHelper(root->left, a);
            if (a.size() > 0) {
                if (a[a.size() - 1] >= root->val)
                    return false;
                else
                    a.push_back(root->val);
            } else {
                a.push_back(root->val);
            }
            bool right = isValidBSTHelper(root->right, a);
            return left && right;
        }
        return true;
    }
};